src/Eval.hs

```diff
@@ -59,23 +59,34 @@ evalStatic ctx xobj = eval ctx xobj PreferGlobal
 -- it gets real weird with laziness. (Note to the note: this code is mostly a
 -- remnant of us using StateT, and might not be necessary anymore since we
 -- switched to more explicit state-passing.)
-eval :: Context -> XObj -> LookupPreference -> IO (Context, Either EvalError XObj)
-eval ctx xobj@(XObj o info ty) preference =
+eval :: Context -> XObj -> LookupPreference -> Bool -> IO (Context, Either EvalError XObj)
+eval ctx xobj@(XObj o info ty) preference shouldResolve =
   case o of
     Lst body -> eval' body
     Sym spath@(SymPath p n) _ ->
-      pure $
-        fromMaybe
-          (evalError ctx ("Can't find symbol '" ++ show n ++ "'") info) -- all else failed, error.
-          -- Certain contexts prefer looking up bindings in the dynamic environment (e.g. defdyanmic) while others
-          -- prefer the static global environment.
-          ( ( case preference of
+      if shouldResolve
+      then
+        pure $
+          case tryAllLookups of
```

Comment:
I would still prefer stylistically to have this check as a call in the monad instead of an explicit case statement; should we need to change where this happens it's a lot more arduous to move case statements around than to swap out function calls in a monad:

```haskell
if shouldResolve 
then unwrapLookup (tryAllLookups >>= checkStatic)
else unwrapLookup tryAllLookups
....
where checkStatic :: (Context, Either EvalError XObj) -> Maybe (Context, Either EvalError XObj)
checkStatic ctx (Right (XObj....))= pure (ctx, Left HasStatic...)
checkStatic ctx val = pure ctx val
```

Reply:
Or, maybe better:

```
(tryAllLookups >>= pure . checkStatic)
```

Reply:
That looks great, I like it! Will update.